util/reflect.go

```diff
@@ -172,7 +172,9 @@ func IsStructValueWithNFields(v reflect.Value, n int) bool {
 
 // InsertIntoSlice inserts value into parent which must be a slice ptr.
 func InsertIntoSlice(parentSlice interface{}, value interface{}) error {
-	DbgPrint("InsertIntoSlice into parent type %T with value %v, type %T", parentSlice, ValueStrDebug(value), value)
+	if DebugLibraryEnabled() {
+		DbgPrint("InsertIntoSlice into parent type %T with value %v, type %T", parentSlice, ValueStrDebug(value), value)
+	}
```

Comment:
How does this help during compilation? Doesn't this mean we have two inline another function during compiler analysis? Do you have evidence that this saves resources during compilation?

Reply:
Good question. From what was observed in profiling, the `DbgPrint` calls were hot paths while the default value of `debugLibrary` in `debug.go` was `false`. It's particularly easy to see the high usages in the flame graph when pretty printing is called inline (and with a large config tree on a resource constrained device). It doesn't seem like the compiler will optimize away these calls when `debugLibrary` is false, possibly because it's a variable that can be changed at runtime.

Reply:
BTW this will *not* save resources at compile time but will noticeably improve runtime performance. I loosely remember this was the first that got patched, and there was a ~75% runtime cut (measured in our integration scenario, YANG models and config data).

Reply:
My guess is this improves performance the same way glog/klog's `.V` works. This looks fine to me.

I'm wondering if you tried adding a wrapper function that does this instead of having to call an if every time?

So something like
```go
// DbgPrint prints v if the package global variable debugLibrary is set.                                                                                                                                                                                                                              
// v has the same format as Printf. A trailing newline is added to the output.                                                                                                                                                                                                                        
func DbgPrint(v ...interface{}) {                                                                                                                                                                                                                                                                     
        if debugLibrary {                                                                                                                                                                                                                                                                             
                dbgPrint(v...)                                                                                                                                                                                                                                                                        
        }                                                                                                                                                                                                                                                                                             
}                                                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                      
// dbgPrint prints v if the package global variable debugLibrary is set.                                                                                                                                                                                                                              
// v has the same format as Printf. A trailing newline is added to the output.                                                                                                                                                                                                                        
func dbgPrint(v ...interface{}) {                                                                                                                                                                                                                                                                     
        out := fmt.Sprintf(v[0].(string), v[1:]...)                                                                                                                                                                                                                                                   
        if len(out) > maxCharsPerLine {                                                                                                                                                                                                                                                               
                out = out[:maxCharsPerLine]                                                                                                                                                                                                                                                           
        }                                                                                                                                                                                                                                                                                             
        fmt.Println(globalIndent + out)                                                                                                                                                                                                                                                               
}                                                                                                                                                                                                                                                                                                     
```
so it's easier for the compiler to inline the first function and then hopefully optimize the rest away.

But if you didn't try this I think we can still go with the current approach. I don't see `DbgPrint` getting heavy usage in the future as glog/klog should be the preferred logging library.

Reply:
The wrapper function works similarly to the existing implementation before this PR. The performance issue is caused by passing the function return as an argument directly to the debug call (maybe I shouldn't have used the `inline` word, it's different than compiler inlining, sorry for the confusion). This results in the argument function call being executed anyway regardless of the `degubLibary` value:

```go
	DbgPrint("InsertIntoMap into parent type %T with key %v(%T) value \n%s\n (%T)",
		parentMap, ValueStrDebug(key), key, pretty.Sprint(value), value)
```

Here the `pretty.Sprint` is computed anyway and is quite expensive.

In `glog` they implement logging calls on a `boolean` type alias and it's an if-statement too. It looks nicer but again if a function return is used as an argument, the `glog` implementation won't prevent unnecessary computations of such functions too, because the computations happen before the if-statement:

```go
// Infof is equivalent to the global Infof function, guarded by the value of v.
// See the documentation of V for usage.
func (v Verbose) Infof(format string, args ...any) {
	if v {
		logf(1, logsink.Info, true, noStack, format, args...)
	}
}
```

Reply:
Thanks for the insight! This looks good.